crates/task-impls/src/quorum_vote.rs

```diff
@@ -0,0 +1,335 @@
+    task::{Task, TaskState},
+};
+use hotshot_types::{
+    event::{Event, EventType},
+    traits::{
+        network::{ConnectedNetwork, ConsensusIntentEvent},
+        node_implementation::{ConsensusTime, NodeImplementation, NodeType},
+    },
+    vote::HasViewNumber,
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+#[cfg(async_executor_impl = "tokio")]
+use tokio::task::JoinHandle;
+use tracing::warn;
+use tracing::{debug, instrument};
+
+/// Vote dependency types.
+#[derive(PartialEq)]
+enum VoteDependency {
+    /// For the `QuorumProposalRecv` event.
+    QuorumProposal,
+    /// For the `DACRecv` event.
+    Dac,
+    /// For the `VidDisperseRecv` event.
+    Vid,
+}
+
+/// Validate the quorum proposal.
+// TODO: Complete the dependency implementation.
+// <https://github.com/EspressoSystems/HotShot/issues/2710>
+#[allow(clippy::needless_pass_by_value)]
+fn validate_proposal<TYPES: NodeType>(_event_sender: Sender<HotShotEvent<TYPES>>) -> bool {
+    true
+}
+
+/// Validate the DAC.
+// TODO: Complete the dependency implementation.
+// <https://github.com/EspressoSystems/HotShot/issues/2710>
+#[allow(clippy::needless_pass_by_value)]
+fn validate_dac<TYPES: NodeType>(_event_sender: Sender<HotShotEvent<TYPES>>) -> bool {
+    true
+}
+
+/// Validate the VID share.
+// TODO: Complete the dependency implementation.
+// <https://github.com/EspressoSystems/HotShot/issues/2710>
+#[allow(clippy::needless_pass_by_value)]
+fn validate_vid<TYPES: NodeType>(_event_sender: Sender<HotShotEvent<TYPES>>) -> bool {
+    true
+}
+
+/// The state for the quorum vote task.
+///
+/// Contains all of the information for the quorum vote.
+pub struct QuorumVoteTaskState<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    /// Smallest potential view number that will be voted for.
+    pub next_vote_view: TYPES::Time,
+
+    /// Table for the in-progress vote dependencies.
+    pub vote_dependencies: HashMap<TYPES::Time, JoinHandle<()>>,
+
+    /// Event receiver.
+    pub event_receiver: Receiver<HotShotEvent<TYPES>>,
+
+    /// Network for all nodes
+    pub quorum_network: Arc<I::QuorumNetwork>,
+
+    /// Network for DA committee
+    pub committee_network: Arc<I::CommitteeNetwork>,
+
+    /// Output events to application
+    pub output_event_stream: async_broadcast::Sender<Event<TYPES>>,
+
+    /// The node's id
+    pub id: u64,
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> QuorumVoteTaskState<TYPES, I> {
+    /// Create an event dependency.
+    #[instrument(skip_all, fields(id = self.id, next_vote_view = *self.next_vote_view), name = "Quorum vote validate VID", level = "error")]
+    fn create_event_dependency(
+        &self,
+        dependency_type: VoteDependency,
+        view_number: TYPES::Time,
+        event_sender: Sender<HotShotEvent<TYPES>>,
+    ) -> EventDependency<HotShotEvent<TYPES>> {
+        EventDependency::new(
+            self.event_receiver.clone(),
+            Box::new(move |event| match dependency_type {
```

Comment:
A potential problem for the next PR:

There might be lifetime issues if passing `&self` inside this `Box`, but we need to update some fields of `self` when verifying the event dependencies.

Reply:
This may work if we update the fields in `handle` where we have `&mut self`, rather than in `create_event_dependency` where accessing `&self` will cause `borrowed data escapes outside of method`.

I can test this idea in the next PR when adding dependency verification logic, but would like to avoid obvious flaws. Curious if anyone has insights on this!

Reply:
Can we pass internal state through a locked arc pointer to avoid the escape issue?

Reply:
I think clone/copy makes sense here. According to the current consensus task implementation where we access `self`, it seems unnecessary to use a lock.